File: c++/encryptor.cpp
//
//  encryptor.cpp
//  Encryptor
//
//  Created by Clifford Panos on 6/10/16.
//  Copyright © 2016 panOS. All rights reserved.
//
//  Will encrypt and decrypt a set of text using a password
//  Uses simple one-character-to-one-character substitution keys
//
//  However, letter-frequency analysis is (somewhat) prevented because more than one key is used to
//  encrypt the characters in the text, and the use of the different keys is dependent on the length
//  of the password. Thus, the cipher text is severely more convoluted to those who do not know the pw.
//

#include <iostream>
#include <deque>
#include <stdio.h>
#include <math.h>
#include <string>
#include <sstream>
using namespace std;


string keyGen(string password) {
    //Generate the initial key using the password
    string stringOfAllCharacters = "&*()-_+=|ABCDEFGHIJK'~,.?<>LMNOPQRSTUVWXYZ/:;[]{}abcdefghijklmnop!@#$%^qrstuvwxyz0123456789";
    stringOfAllCharacters.push_back('"');
    //stringOfAllCharacters replaced the space character with a tilde here to prevent user input error
    string key = password; //Places the password at the beginning of the key
    bool isTheCharacterAlreadyInThePassword;
    for (unsigned int i=0; i <= stringOfAllCharacters.length() - 1; i++) {
        isTheCharacterAlreadyInThePassword = false;
        for (unsigned int j=0; j <= password.length() - 1; j++) {
            if (stringOfAllCharacters[i] == password[j])
                isTheCharacterAlreadyInThePassword = true;
        }
        if (isTheCharacterAlreadyInThePassword == false)
            key += stringOfAllCharacters[i];
    }
    return key;
}

string extraKeyGen(string password, string initialKey, int keyNum) {
    //Generate the second, third, fourth, and/or fifth keys to be used
    //These extra keys are used to make letter-frequency analysis nearly impossible

    string newKey = initialKey; //This initialKey will now be modified to produce the newKey
    string sectionOfPasswordToBeTranslated = password.substr((2*keyNum - 2), int(password.length() - 1));
    newKey.insert(int(newKey.length() - sectionOfPasswordToBeTranslated.length()), sectionOfPasswordToBeTranslated);
    newKey.erase((2*keyNum - 2),sectionOfPasswordToBeTranslated.length());
    if (keyNum % 2 == 0 && password.length() % 2 == 0)
        newKey = newKey.substr(45, newKey.length() - 1) + newKey.substr(0,45);
    if (keyNum % 2 == 1 && password.length() % 2 == 1)
        newKey = newKey.substr(30, 20) + newKey.substr(0,30) + newKey.substr(75, newKey.length() - 1) + newKey.substr(50,25);
    //cout << "newKey " << keyNum << " is: " << '"' << newKey << '"' << endl;
    return newKey;
}

int main() {

    string START = "yes"; // The body of this function takes place in a do-while loop;

    do {

        string input;
        string output = "";
        string keyword, password, text;
        string stringOfAcceptablePasswordCharacters = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz?!@#$%^&*(){}[]<>:;/._-=+";
        string stringOfAllCharacters = "' ,.?<>/:;[]{}!@#$%^&*()-_+=|ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
        stringOfAllCharacters.push_back('"');

        cout << "Enter the password & text to be encrypted or decrypted in the following format:" << endl;
        cout << "<(encrypt or decrypt)>, <password>, <text>" << endl;
        getline(cin, input);
        if (input.length() < 7) {
            cout << endl << "Please enter three separate arguments. Separate each argument with a comma." << endl;
            START = "yes";
            continue;
        }

        //Remove spaces (' ') between the three arguments
        string temporaryProcessedInput = "";
        int numCommasReached = 0;
        int b = 0; //iterator for the while loop
        while (numCommasReached <2) {
            if (input[b] == ',')
                numCommasReached++;
            if (input[b] != ' ')
                temporaryProcessedInput += input[b];
            b++;
        }
        for (unsigned int finalSpaceIterator = b; finalSpaceIterator <= input.length() - 1; finalSpaceIterator++) {
            if (input[finalSpaceIterator] == ' ')
                b++;
            else break;
        }
        temporaryProcessedInput += input.substr(b, int(input.length() - 1));
        input = temporaryProcessedInput; //This new input now has no spaces between arguments



    //Now the program will check for all possible User Input Errors



        //Checks to make sure two commas were entered and locates the iterative places of those commas
        int numCommas = 0; //How many commas have been encountered at the iterative moment
        int commaOne = 0; //iterative comma place one
        int commaTwo = 0; //iterative comma place two
        int checkIfThereIsntAThirdArgument = 0;
        for (unsigned int a=0; a <= input.length() - 1; a++) {
            checkIfThereIsntAThirdArgument = a;
            if (numCommas == 2)
                break;
            if (input[a] == ',') {
                if (numCommas == 0)
                    commaOne = a;
                if (numCommas == 1)
                    commaTwo = a;
                numCommas++;
            }
        }
        // Checks to make sure enough arguments were entered
        if ( (numCommas < 2) || (input.length() <= int(checkIfThereIsntAThirdArgument + 1)) ) {
            if (input.length() <= int(checkIfThereIsntAThirdArgument + 1))
                cout << endl << "You did not enter text to be used; please start again and enter some text as the third argument." << endl;
            else cout << endl << "Please enter three separate arguments in the format:" << endl << "<(encrypt or decrypt)>, <password>, <text>" << endl << "Separate each argument with a comma." << endl;
            START = "yes";
            continue;
        }


        /*Iterrupt process of checking User Inpupt Error in order to separate the three
         arguments of the input into three separate variables. i.e. Send the first argument to
         the variable 'keyword', send the second argument to 'password', and the third variable
         to 'text'. */
        //str::substr(arg) is arg = initial position, length of span
        keyword = input.substr(0,int(commaOne));
        password = input.substr(int(commaOne+1),int(commaTwo - commaOne - 1));
        text = input.substr(int(commaTwo+1),int(input.length()-1));



        //Ensure that the first word is either encrypt or decrypt
        bool didTheUserChooseEncryptOrDecrypt = false;
        if (keyword == "Encrypt" || keyword == "Enc" || keyword == "encrypt" || keyword == "enc" || keyword == "ENCRYPT" || keyword == "ENC") {
            didTheUserChooseEncryptOrDecrypt = true;
            keyword = "encrypt";
        }
        if (keyword == "Decrypt" || keyword == "Dec" || keyword == "decrypt" || keyword == "dec" || keyword == "DECRYPT" || keyword == "DEC") {
            didTheUserChooseEncryptOrDecrypt = true;
            keyword = "decrypt";
        }
        if (didTheUserChooseEncryptOrDecrypt == false) {
            cout << endl << "Please start again and choose the first word to be either encrypt or decrypt." << endl;
            START = "yes";
            continue;
        }


        //Ensure that the password contains only acceptablePasswordCharacters
        if (password.length() == 0) {
            START = "yes";
            cout << endl << "You did not enter a password; please start again using a password." << endl;
            continue;
        }
        int numberOfCharactersFoundThatAreAcceptable = 0;
        for (unsigned int w = 0; w <= password.length() - 1; w++) {
            for (unsigned int x = 0; x <= stringOfAcceptablePasswordCharacters.length() - 1; x++) {
                if (password[w] == stringOfAcceptablePasswordCharacters[x])
                    numberOfCharactersFoundThatAreAcceptable += 1;
            }
        }
        if (numberOfCharactersFoundThatAreAcceptable < password.length()) {
            START = "yes";
            cout << endl << "Acceptable characters for a password are A-Z, a-z, 0-9, and ?!@#$%^&*(){}[]<>:;/,._-=+" << endl;
            cout << "Please start again using a different password." << endl;
            continue;
        }
        //DELETES DUPLICATE CHARACTERS IN THE PASSWORD; i.e. "hashbrowns has two 's' and two 'h'"
        string duplicatePassword = ""; //This string will be the duplicate of the password and will be used to delete repeats
        duplicatePassword += password[0]; //The first character of password is added to this no matter what
        bool duplicate = false;			//This bool stores whether or not the character is duplicated
        for (int f=0; f <= password.length() - 1; f++) {    //This will actually duplicate possibleSolutions into duplicateList
            for (int g=0; g <= duplicatePassword.length() - 1; g++) {
                if (password[f] == duplicatePassword[g])
                    duplicate = true;
            }
            if (duplicate == false)  //If the number makes it through the whole loop without being found to be a repeat
                duplicatePassword += password[f];
            else duplicate = false;
        }
        password = duplicatePassword;


        //Generate between 1 and 5 keys depending on the length of the password
        int numberOfKeysToUse = 0;
        string key2, key3, key4, key5;
        if (password.length() == 1 || password.length() == 2) numberOfKeysToUse = 1;
        if (password.length() == 3 || password.length() == 4) numberOfKeysToUse = 2;
        if (password.length() == 5 || password.length() == 6) numberOfKeysToUse = 3;
        if (password.length() == 7 || password.length() == 8) numberOfKeysToUse = 4;
        if (password.length() >= 9) numberOfKeysToUse = 5;
        if (password.length() >= 9 && password.length() % 2 == 0) numberOfKeysToUse = 4;
        if (password.length() >= 9 && password.length() % 3 == 0) numberOfKeysToUse = 3;

        string initialKey = keyGen(password); //Generates the initialKey
        if (numberOfKeysToUse > 1) { //Produces the extra keys if the password is long enough
            for (unsigned int c=2; c <= numberOfKeysToUse; c++) {
                if (c == 2) key2 = extraKeyGen(password, initialKey, 2);
                if (c == 3) key3 = extraKeyGen(password, initialKey, 3);
                if (c == 4) key4 = extraKeyGen(password, initialKey, 4);
                if (c == 5) key5 = extraKeyGen(password, initialKey, 5);
            }
        }


        if (keyword == "encrypt") {
            /* If a text character entered is not in stringOfAllCharacters, it is
            substituted for a question mark '?' character */
            bool isThisAnUnacceptableTextCharacter;
            numberOfCharactersFoundThatAreAcceptable = 0; //This is the same variable as used to determine if the password characters were acceptable
            for (unsigned int y = 0; y <= text.length() - 1; y++) {
                isThisAnUnacceptableTextCharacter = true;
                for (unsigned int z = 0; z <= stringOfAllCharacters.length() - 1; z++) {
                    if (text[y] == stringOfAllCharacters[z]) {
                        numberOfCharactersFoundThatAreAcceptable += 1;
                        isThisAnUnacceptableTextCharacter = false;
                    }
                }
                if (isThisAnUnacceptableTextCharacter == true) {
                    cout << "unacceptable at: " << y << endl;
                    text.replace(y, 1, "?");
                }
            }
            if (numberOfCharactersFoundThatAreAcceptable < text.length()) cout << endl << "One or more text characters cannot be encrypted and will instead be decrypted as '?' signs." << endl;

            //The loop below will encrypt the message using the initialKey and any extraKeys
            for (unsigned int q = 0; q <= text.length() - 1; q+=numberOfKeysToUse) {
                if (numberOfKeysToUse > 0 && ((q+0) <= (text.length() - 1)))
                    output += initialKey[stringOfAllCharacters.find(text[q])];
                if (numberOfKeysToUse > 1 && ((q+1) <= (text.length() - 1)))
                    output += key2[stringOfAllCharacters.find(text[q+1])];
                if (numberOfKeysToUse > 2 && ((q+2) <= (text.length() - 1)))
                    output += key3[stringOfAllCharacters.find(text[q+2])];
                if (numberOfKeysToUse > 3 && ((q+3) <= (text.length() - 1)))
                    output += key4[stringOfAllCharacters.find(text[q+3])];
                if (numberOfKeysToUse > 4 && ((q+4) <= (text.length() - 1)))
                    output += key5[stringOfAllCharacters.find(text[q+4])];
            }

            //cout << "keyword is: " <<  '"' << keyword <<  '"' << endl;
            //cout << "password is: " <<  '"' << password <<  '"' << endl;
            //cout << "text is: " <<  '"' << text <<  '"' << endl;
            cout << endl << "The encrypted cipher text is: " << endl << endl << output << endl << endl;
        }

        if (keyword == "decrypt") {
            //The decrypt function does not need to modify the input text because it comes from the above encrypt function
            //The loops below will decrypt the message using the initialKey and any extraKeys
            for (unsigned int r = 0; r <= text.length() - 1; r+=numberOfKeysToUse) {
                if (numberOfKeysToUse > 0 && ((r+0) <= (text.length() - 1)))
                    output += stringOfAllCharacters[initialKey.find(text[r])];
                if (numberOfKeysToUse > 1 && ((r+1) <= (text.length() - 1)))
                    output += stringOfAllCharacters[key2.find(text[r+1])];
                if (numberOfKeysToUse > 2 && ((r+2) <= (text.length() - 1)))
                    output += stringOfAllCharacters[key3.find(text[r+2])];
                if (numberOfKeysToUse > 3 && ((r+3) <= (text.length() - 1)))
                    output += stringOfAllCharacters[key4.find(text[r+3])];
                if (numberOfKeysToUse > 4 && ((r+4) <= (text.length() - 1)))
                    output += stringOfAllCharacters[key5.find(text[r+4])];
            }
            cout << endl << "Originally encrypted text characters that were unacceptable have been decrypted as '?' signs." << endl;
            cout << endl << "The decrypted text is: " << endl << endl << output << endl << endl;
        }

        cout << "Would you like to restart this function and use it again?: ";
        getline(cin, START);
        if (START == "yes" || START == "YES" || START == "Yes" || START == "yes " || START == " yes " || START == "yup" || START == "sure")
            START = "yes";
        cout << endl << endl;

    } while (START == "yes"); //Ends the do-while loop

    return 0; //Ends int main()
}

File: src/controller/Encryptor.java
package controller;

//
//  Encryptor
//  Encryptor.java
//
//  Created by Clifford Panos and edited with Anthony D'Achille on 6/10/16.
//  Copyright © 2016 panOS. All rights reserved.
//
//  Will encrypt and decrypt a set of text using a password.
//  Uses multiple one-character-to-one-character substitution keys.
//  A third party can have access to this code and still not be able to decrypt.
//  Similar passwords to the encryptor's password will not work.
//
//  However, letter-frequency analysis is [somewhat] prevented because more than
//  one key is used to encrypt the characters in the text, and the use of the
//  different keys is dependent on the length of the password. Thus, the cipher
//  text is severely more convoluted to those who do not know the pw.
//
//  Additionally, the text is encrypted and then re-encrypted recursively.
//

import view.UIAlert;
import view.EncryptDecryptMenu;
import view.InputOnEncryptMenu;
import view.MainScreen;

import java.io.File;


public class Encryptor {

    // Class-wide variables
    private static String allCharactersWithSpace =
        " &*()-_+=|ABCDEFGHIJK',.?<>LMNOPQRS"
        + "TUVWXYZ/:;[]{}abcdefghijklmnop!@#$%^qrstuvwxyz0123456789\"";
        /* allCharacters replaced the space character with a tilde here to
           prevent user input error */

    private static String keyword = null;
    private static String password = null;
    private static String text = null;
    private static String passwordHashAlpha;
    private static String passwordHashBeta;
    private static int numKeys = 0;

    private static String pwCheck = "$Enc$"; //See below for explanation
    private static String nLSubstitute = "/N/$/L/";
        //Used to substitute newLine characters


    public static void setKeyword(String aKeyword) {
        keyword = aKeyword;
    }

    public static void setPassword(String aPassword) {
        password = aPassword;
    }

    public static void setText(String aText) {
        text = aText;
    }

    public static String run() {

        // GET THE KEYWORD FROM THE USER
        if (keyword == null) {
            UIAlert.show("Choose to Encrypt or Decrypt",
                "You did not choose to encrypt or decrypt.\n"
                + "Please choose from the leftmost pane.",
                javafx.scene.control.Alert.AlertType.ERROR);
            return null;
        }

        // GET THE PASSWORD FROM THE USER
        password = EncryptDecryptMenu.getPasswordFieldText();
        if (password == null || password.equals("")) {

            UIAlert.show("Enter a Password",
                "You did not enter a password with\n"
                + "which to " + keyword + " the text. Please\n"
                + "enter a password in the leftmost pane.",
                javafx.scene.control.Alert.AlertType.ERROR);
            return null;
        }
        //ELSE:
        int acceptablePwCharactersFound;
        String acceptablePwCharacters = "0123456789ABCDEFGHIJKLMNOPQRSTUVWX"
            + "YZabcdefghijklmnopqrstuvwxyz?!@#$%^&*(){}[]<>:;/._-=+";

        //Ensure that the pw contains only acceptablePwCharacters
        acceptablePwCharactersFound = 0;
        for (int i = 0; i < password.length(); i++) {
            if (acceptablePwCharacters.indexOf(password.charAt(i)) != -1) {
                acceptablePwCharactersFound += 1;
            }
        }

        if (acceptablePwCharactersFound < password.length()) {
            UIAlert.show("Unacceptable Password",
                "You entered a password with invalid characters.\n"
                + "Acceptable characters for a password are:\n"
                + "A-Z, a-z, 0-9, and ?!@#$%^&*(){}[]<>:;/,._-=+\n"
                + "Please enter an acceptable password & try again.",
                javafx.scene.control.Alert.AlertType.ERROR);
            return null;
        }

        // Special operations for getting text variable if inputting directly
        if (MainScreen.isInputtingDirectly()) {
            // GET THE TEXT TO BE ENCRYPTED/DECRYPTED FROM THE USER
            if (MainScreen.getIsEncryptingNotDecrypting()) {
                text = MainScreen.getInputOEMenu().getInputFieldText();
            } else {
                text = MainScreen.getInputODMenu().getInputFieldText();
            }

            if (text == null || text.equals("")) {

                UIAlert.show("Enter Text to be " + keyword + "ed",
                    "You did not input any text to be " + keyword + "ed.\n"
                    + "Please type or copy text into the box.",
                    javafx.scene.control.Alert.AlertType.ERROR);
                return null;
            }
        }

        System.out.println("Text is: '" + text + "'");



        //At this point, the static variables keyword, password, and text have
        //their respective values assigned to them & are ready to be used.



        //GENERATE initialKey
            // ---> eliminates repeats from password & allCharacters
        String initialKey = KeySet.keyGen(password);

        //We already have the password, so generate the Beta Password:
        String betaPassword = "";
        for (int p = password.length() - 1; p >= 0; p--) {
            betaPassword += password.charAt(p);
        }

        //GENERATE PASSWORDHASHALPHA and PASSWORDHASHBETA ----------------
        passwordHashAlpha = hashThePassword(password);
        passwordHashBeta = hashThePassword(betaPassword);

        //Now, IndexCurve and Recursion can be constructed properly
        IndexCurve indexCurve =
            new IndexCurve(passwordHashAlpha, passwordHashBeta);

        //Assign the correct value to numKeys
        numKeys =
            indexCurve.curve((password.length() * password.length()), 6, 18);
        Recursion.setNumKeys(numKeys);



// These print statements are for testing
        System.out.println("password: '" + password + "'");

        if (keyword.equals("encrypt")) {
            return encrypt();
        } else if (keyword.equals("decrypt")) {
            return decrypt();
        } else {
            return null;
        }

    }

    //ENCRYPT

    private static String encrypt() {

        if (hasEncryptedTags(text)) {
            UIAlert.show("Text Already Encrypted",
                "The text that you entered has already\n"
                + "been encrypted. To prevent the convolution\n"
                + "inherent in multiple encryptions, this \n"
                + "text will not been encrypted again.",
                javafx.scene.control.Alert.AlertType.ERROR);
            return null;
        }

        String temporaryText = "";
        for (char c : text.toCharArray()) {
            if (!String.valueOf(c).matches(".")) {
                System.out.println("NOT A MATCH");
                temporaryText += nLSubstitute;
            } else {
                temporaryText += c;
            }
        }
        text = temporaryText;

        //will hold all user-entered characters that are invalid
        String invalidCharacterList = "";
        String textNoSpaces = "";
        for (char c : text.toCharArray()) {
            textNoSpaces += (c == ' ' ? "~"
                : (allCharactersWithSpace.indexOf(c) == -1 ? "?" : c));
            //If a character is a space, make it a ~ so that it will work
                //in Recursion's encrypt function
            //If the character is not in allCharactersWithSpace, then change
                // it to be a ? so that it will work in Recursion's encrypt
            invalidCharacterList +=
                (allCharactersWithSpace.indexOf(c) == -1
                    ? (invalidCharacterList.indexOf(c) == -1 ? c : "")
                    : "");
            //If the character IS invalid AND it is not already in
                //invalidCharacterList, then add it to invalidCharacterList
        }

        text = pwCheck + textNoSpaces;
        /*The word Encryptor is added so that it can be checked if it
          decrypted properly when the user tries to decrypt. If on
          decryption, the first 9 characters are not '$Enc$', then
          the user entered an incorrect password, and the remaining text
          will not be decrypted because doing so would harm the data.*/
        System.out.println("New text is: " + text);

        String finalEncryptedCipherText = Recursion.encrypt(text, numKeys);
        //encrypt() works from numKeys down to 1 by decrementing

        String invalidPrintList = "";
        for (char c : invalidCharacterList.toCharArray()) {
            invalidPrintList += (c + ", ");
        }
        if (invalidPrintList.length() > 1) {
            invalidCharacterList = invalidPrintList
                .substring(0, invalidPrintList.length() - 2); //remove space
            System.out.println("Invalid charList: " + invalidCharacterList);

        }

        if (invalidCharacterList.length() > 0) {
            boolean cancelling = UIAlert.show("Invalid Characters Entered",
                "The following characters are invalid:\n  "
                + invalidCharacterList + "\n\n"
                + "When these characters are decrypted,\n"
                + "they will appear as '?' characters.\n\n"
                + "Would you like to proceed?",
                javafx.scene.control.Alert.AlertType.CONFIRMATION, true);

            if (cancelling) {
                return null; //Exit the encryption process
            }
        }

        //Add two %E% & $E$ tags to prevent multiple encryptions:
        finalEncryptedCipherText
            = "%E%" + finalEncryptedCipherText + "$E$";

        System.out.println("\n\nEncrypted text is: "
            + finalEncryptedCipherText);
        return finalEncryptedCipherText;

    }

    //DECRYPT

    private static String decrypt() {

        //Test to make sure that decrypting the first nine characters
        //returns '$Enc$'
        //If it does not, decryption will halt.
        if (text.length() < 11) {
            UIAlert.show("Inputted Text Incorrect",
                "The text that you entered to decrypt is incorrect.\n"
                + "All correctly inputted text is greater than length 11.",
                javafx.scene.control.Alert.AlertType.ERROR);
            return null;
        }

        if (!hasEncryptedTags(text)) {
            UIAlert.show("Text Not Decryptable",
                "The text that you are attempting to \n"
                + "decrypt has NOT been encrypted by this system.\n"
                + "To prevent a loss of data through false\n"
                + "decryption, this text will not be decrypted.",
                javafx.scene.control.Alert.AlertType.ERROR);
            return null;
        }

        //Remove the %E% & $E$ tags that were added at the end of the encryption
        text = text.substring(3, text.length() - 3);

        if (!(Recursion.decrypt(text.substring(0, 5), 1)).equals(pwCheck)) {
            UIAlert.show("Incorrect Password",
                "The decryption password that you entered is incorrect.\n"
                + "Attempting to decrypt the text with an incorrect\n"
                + "password would render a disarray of useless data.\n\n"
                + "Try to decrypt again using a different password.",
                javafx.scene.control.Alert.AlertType.ERROR);
            return null;
        }
        String finalDecryptedCipherText =
            Recursion.decrypt(text, 1);
            //The second argument of decrypt() MUST be 1 because the
            //encrypt function works from 1 to numKeys
            //this is because it must undo the encrypt function oppositely
        String decryptedNoTildes = "";
        for (char c : finalDecryptedCipherText.toCharArray()) {
            decryptedNoTildes += (c == '~' ? " " : c);
            //Spaces were made into ~ characters before being encrypted, so
                //now they must be transformed back into spaces.
        }
        finalDecryptedCipherText = decryptedNoTildes.substring(5);
            //This goes from index 9 to the end to remove the "$Enc$"


        String temporaryText = "";
        for (int i = 0; i < finalDecryptedCipherText.length(); i++) {
            if (!(i > finalDecryptedCipherText.length() - 7)
                && finalDecryptedCipherText.substring(i, i + 7)
                    .equals(nLSubstitute)) {
                System.out.println("NEWLINE DECRYPTING");
                temporaryText += "\r\n";
                i += 6; //the for loop already increments i by 1
            } else {
                temporaryText += finalDecryptedCipherText.charAt(i);
            }
        }
        finalDecryptedCipherText = temporaryText;
        System.out.println("\nDecrypted text is: "
            + finalDecryptedCipherText);
        return finalDecryptedCipherText;
    }



    //--------------------------------------------------------------------//



    public static String hashThePassword(String pwToHash) {

        //Custom hashCode function to provide cross-platform / language support
        long longHash = 0;
        int off = 0;
        char[] values = pwToHash.toCharArray();
        int len = pwToHash.length();
        for (int i = 0; i < pwToHash.length(); i++) {
            longHash = 31 * longHash + Character.getNumericValue(values[off++]);
        }

        if (longHash < 0) {
            longHash = longHash * -1; //Ensure that longHash is positive
        }

        String stringOfTheHash = "" + longHash;

        //Eliminate zeroes from the password hashes
        for (int i = 0; i < stringOfTheHash.length(); i++) {
            if ((stringOfTheHash.charAt(i)) == ('0')) {
                stringOfTheHash = stringOfTheHash.substring(0, i)
                    + i + stringOfTheHash.substring(i + 1);
            }
        }
        return stringOfTheHash;

    }

    public static String getPasswordHashAlpha() {
        return passwordHashAlpha;
    }

    public static String getPasswordHashBeta() {
        return passwordHashBeta;
    }

    public static int getNumKeys() {
        return numKeys;
    }

    public static boolean hasEncryptedTags(String textToCheck) {
        boolean hasTheTags = false;
        if (textToCheck != null && textToCheck.length() >= 6) {
            System.out.println("TTT: '" + textToCheck + "'");
            System.out.println(textToCheck.substring(0, 3));
            System.out.println(textToCheck.substring(textToCheck.length() - 3));
            if (textToCheck.substring(0, 3).equals("%E%")
                && textToCheck.substring(textToCheck.length() - 3)
                    .equals("$E$")) {
                hasTheTags = true;
            }
        }
        return hasTheTags;
    }

}
